plugin/auth/oauth.go

```diff
@@ -0,0 +1,159 @@
+package auth
+
+// FLOW
+// if not authenticated
+// - button in web settings
+// - redirect to oauth2 provider
+// - redirect back to evcc with code
+// - state authenticated
+// if authenticated
+// - refresh token on demand
+// - save token to db regularly
+
+// STATES
+// not authenticated
+// authenticated
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/server/authredirect"
+	"github.com/evcc-io/evcc/server/db/settings"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"golang.org/x/oauth2"
+)
+
+type OAuth struct {
+	mu      sync.Mutex
+	cc      oauth2.Config
+	subject string
+	ts      oauth2.TokenSource
+	cv      string
+	log     *util.Logger
+}
+
+var (
+	mu         sync.Mutex
```

Comment:
```suggestion
	oauthMu         sync.Mutex
```

...das Package enthält ja noch andere Files